Implement metrics for cache hit ratio

Robin-07 opened the pull request:
Fixes #3538 

## Summary
Added `cache_hits` and `cache_misses` metrics to calculate cache hit ratio.

## Changes
- Add instrumentation to `cache_backend.get()` by wrapping it on application startup.
- Add tests.

## Checklist

- [x] Add tests.
- [x] Add your name in the contributors file.
- [x] Add documentation.
- [ ] Add a changelog entry.

Robin-07 commented:
Thanks for your feedback. Yes, the composition pattern totally makes sense. We can add a `metrics` attribute to the `cache`, however, one problem I see with doing it in the `load_from_config(config)` methods is that at that point, `config.registry.metrics` hasn't been set (because the `initialization_sequence` is setting up the `cache` before `metrics`). And changing this order would also break things as `setup_metrics(config)` expects the `cache` to be already set up as below: 

```python
def setup_metrics(config):
---
    def on_app_created(event):
        config = event.app
        metrics_service = config.registry.metrics

        metrics.watch_execution_time(metrics_service, config.registry.cache, prefix="backend")
```

One option is to do `load_from_config(config)` during the initial cache setup and set the `cache.metrics` attribute during the metrics setup later. Let me know if that sounds good.

leplatrem commented:
> One option is to do load_from_config(config) during the initial cache setup and set the cache.metrics attribute during the metrics setup later. Let me know if that sounds good.

Yes, that would be clean indeed.
You could do that in `setup_metrics()` (either through a method `cache_backend.set_metrics()` or directly `cache_backend.metrics = ...`.
I don't know if `metrics` is a good name, maybe? Otherwise something like `metrics_service` or `metrics_backend`.

Robin-07 commented:
@leplatrem Updated the PR with following changes:

1. Added attribute `metrics_backend` and corresponding setter `set_metrics_backend()` to the CacheBase class. 
2. Added a small adapter class `CacheMetricsBackend` for cache-related metrics accounting. An instance of this is what is stored in the `metrics_backend` attribute. Currently it provides 2 utility methods `count_miss()` and `count_hit()`. More methods can be added in future for other metrics.
3. Default `metrics_backend` is set using `NoOpMetricsService` in the `load_from_config()` methods. If metrics are enabled, `cache_backend.set_metrics_backend()` is used to update the chosen backend inside the `setup_metrics()` method.
4. Added test cases for all 3 cache backends.

Please let me know your thoughts.

leplatrem reviewed (changes requested):
(oops, marked as approved but we still have a few more improvements to do before merge)

Robin-07 commented on kinto/core/cache/postgresql/__init__.py:
```diff
@@ -172,4 +177,8 @@ def delete(self, key):
+    return Cache(
+        client=client,
+        cache_prefix=settings["cache_prefix"],
+        metrics_backend=NoOpMetricsService(),
```
Makes sense 👍🏻
